core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -831,6 +835,14 @@ private[spark] class AppStatusListener(
         }
       }
     }
+    // Flush updates if necessary. Executor heartbeat is an event that happens periodically. Flush
+    // here to ensure the staleness of Spark UI doesn't last more that the executor heartbeat
+    // interval.
+    if (now - lastFlushTimeNs > liveUpdatePeriodNs) {
+      flush(maybeUpdate(_, now))
```

Comment:
Hmm... in the bug you mention that job-level data is not being updated. Is that the only case? Because if that's it, then this looks like overkill. You could e.g. update the jobs in the code that handles `event.accumUpdates` above, or even just flush jobs specifically, instead of everything.

Doing a full flush here seems like overkill and a little expensive when you think about many heartbeats arriving in a short period (even when considering `lastFlushTimeNs`).

Reply:
> Hmm... in the bug you mention that job-level data is not being updated. Is that the only case?

I also noticed that executor active tasks sometimes could be wrong. That's why I decided to flush everything to make sure we don't miss any places. It's also hard to maintain if we need to manually flush in every place.

Ideally, we should flush periodically so that it doesn't depend on receiving a Spark event. But then I will need to add a new event type and post to the listener bus. That's overkilled.

> when you think about many heartbeats arriving in a short period

At least there will be at least 100ms between each flush. As long as we process heart beats very fast, most of them won't trigger the flush.

Reply:
If the goal is to use the hearbeats as some trigger for flushing, how about using some ratio of the heartbeat period instead of `liveUpdatePeriodNs` to control whether to flush everything?

Really large apps can get a little backed up when processing hearbeats from lots and lots of busy executors, and this would make it a little worse.

Reply:
The update only happens in live UI, which should be fine in general.  For real large apps, will it help by setting `LIVE_ENTITY_UPDATE_PERIOD` to a larger value? Setting a ratio of heartbeat period seems a bit complex.

Reply:
> only happens in live UI

The "don't write to the store all the time" thing was added specifically to speed up live UIs, because copying + writing the data (even to the memory store) becomes really expensive when you have event storms (think thousands of tasks starting and stopping in a very short period).

>  setting LIVE_ENTITY_UPDATE_PERIOD to a larger value

We should avoid requiring configuration tweaks for things not to break, when possible.